cpp/src/join/chuck_join.cu

```diff
@@ -0,0 +1,114 @@
+#include <cudf/copying.hpp>
+#include <cudf/join.hpp>
+#include <cudf/table/table_view.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/gather.hpp>
+#include <hash/concurrent_unordered_map.cuh>
+
+#include "hash_join.cuh"
+
+namespace cudf {
+
+namespace detail {
+
+std::unique_ptr<cudf::experimental::table> left_semi_anti_join(cudf::table_view const& left,
+                                                               cudf::table_view const& right,
+                                                               std::vector<cudf::size_type> const& left_on,
+                                                               std::vector<cudf::size_type> const& right_on,
+                                                               std::vector<cudf::size_type> const& return_columns,
+                                                               JoinType join_type,
```

Comment:
What if `return_columns` is empty ?

Reply:
A generic c++ stl exception :-)

I'll add a CUDF_EXPECTS at the beginning.

Reply:
Generally in libcudf empty inputs (even index lists) typically result in empty outputs, not exceptions, unless that is not possible. This is because e.g. in Dask partitions can result in empty results for some workers which are then used as inputs.

Reply:
Do we have a mechanism for vertical partitioning of a data frame?  The only way I could see Dask partitions with some partitions having empty return_columns and other partitions having a value in return_columns would be if we did vertical partitioning.

It's simple enough to return a table with no columns if that's the behavior we want.

Reply:
That's generally what we do. And yes, we can vertically partition: https://github.com/rapidsai/cudf/blob/e9b8748fdc3aaacc641a07b07be78c79ef5110dc/cpp/include/cudf/hashing.hpp#L23-L44

Reply:
Unless I'm misunderstanding the description, this appears to do horizontal
partitioning (at least as I was intending in my comment above).

In horizontal partitioning, the table is partitioned so that whole rows stay together and some rows go in Partition A and some rows go in Partition B.

In vertical partitioning, the table is partitioned so that some columns go in partition A and other columns go in partition B.

It looks to me like our hashing code does horizontal partitioning (for each row, create a hash value, use that hash value to decide which partition the row belongs to).

If, in fact, we only do horizontal partitioning, then it's not clear to me why we would end up calling left_semi_join with return_columns empty in one dask partition and with it not empty in some other dask partition.  It seems like that particular parameter should be the same regardless of which dask partition is being accessed.  In this case, an exception would seem reasonable.

Reply:
Sorry, I considered vertical partitioning to mean the split points are along the vertical axis (cutting columns into pieces -- I think of columns as vertical). Terminology disconnect.

I haven't thought about the specifics. I was just saying that in most cases, empty inputs result in empty outputs in our library. If that doesn't make sense for this algorithm for some reason, that's fine. What the algorithm does has escaped my brain since I reviewed.

Reply:
I can see asking for zero columns to be returned if a semi/anti join was coalesced with a subsequent column drop operation where all table columns happened to be dropped.  In Spark, if all the columns of a table are dropped the result is a table that has no columns but still has as many rows as the original table.

So my vote for this behavior would be to return a table with zero columns but the row count set to the number of rows in the left table that survived the join operation.  This could also be useful for the scenario where someone only wants the number of rows that survive the semi/anti join and does not want to manifest any row data.

Reply:
It's not possible to have a table of zero columns with non-zero rows.

Reply:
> It's not possible to have a table of zero columns with non-zero rows.

Is that something to reconsider?  There are other cases where this could be useful for operations that would normally return a full table but the caller is only interested in a count.  For example, asking libcudf to parse an input file but the caller is only interested in the row count and does not want to manifest any row data.

For this PR, I think it needs to throw since its hands are tied and it can't do the right thing.  IMO returning an empty table with no columns and no rows isn't going to be very useful in practice.

Reply:
> Is that something to reconsider?

How can you have "rows" if there aren't any columns to contain the rows? 

It feels like violating a universal axiom. It's like violating `0*x == 0 ∀ x`. At the very least I think it violates the principle of least surprise. 

> caller is only interested in a count

Sounds like a separate function to me, but would depend on the exact use case.

Reply:
To me it's not much more surprising than a table with columns but no rows.  One has a degenerate X dimension, the other a degenerate Y dimension.

If libcudf wants to provide separate APIs for methods to handle obtaining a count without manifesting data that works too.  Tables with rows but no columns allows this to be done without increasing the API footprint and aligns with how some other frameworks handle it.  However I'm more interested in the ability to accomplish the task rather than the syntax for doing so.

Reply:
> To me it's not much more surprising than a table with columns but no rows. One has a degenerate X dimension, the other a degenerate Y dimension.

The difference is it's possible to have a column without any elements. The inverse doesn't hold. It's not possible to have elements without columns.

Reply:
> It's not possible to have elements without columns.

The elements are present, they just have no data associated with them and therefore take no space.  This is akin to a column with no rows being present, but there's no data associated with it and therefore it takes no space.

You can think of a row in a table as being an array of values, where each entry in the array corresponds to a column in the table.  A table would then be an array of rows, or an array of arrays.  When a table has no rows, it's an empty array, i.e.: `[]`.  When a table contains no columns but three rows, it's an array of three empty arrays, i.e.: `[ [], [], [] ]`.

Reply:
This seems like something to address later in a different PR.  At the C++ level, it would be easy to add a counting version (skips the gather call) that computes the counts for any of the join operations.  Spark or Python could wrap them differently if desired.